### ISSUE TITLE
Not able to display participant video

### ISSUE THREAD

**Issue by 12MIT0020** [twilio/twilio-video.js#598]
- [ ] I have verified that the issue occurs with the latest twilio-video.js release and is not marked as a known issue in the [CHANGELOG.md](https://github.com/twilio/twilio-video.js/blob/master/CHANGELOG.md).
 - [ ] I reviewed the [Common Issues](https://github.com/twilio/twilio-video.js/blob/master/COMMON_ISSUES.md) and open GitHub issues and verified that this report represents a potentially new issue.
 - [ ] I verified that the [Quickstart](https://github.com/twilio/video-quickstart-js) application works in my environment.
 - [ ] I am not sharing any [Personally Identifiable Information (PII)](https://www.twilio.com/docs/glossary/what-is-personally-identifiable-information-pii)
       or sensitive account information (API keys, credentials, etc.) when reporting this issue.

**Code to reproduce the issue:**
`
function roomJoined(room) {
    console.log(room);
    window.room = activeRoom = room;
  
    var previewContainer = angular.element('#local-media');
  
      attachParticipantTracks(room.localParticipant, previewContainer[0]);
  
    room.on('participantConnected', function(participant) {
     
    participant.tracks.forEach(function(participant) {
        //log("Already in Room: '" + participant.identity + "'");
        var previewContainerPat = angular.element('#remote-media');
        console.log(previewContainerPat);
        attachParticipantTracks(participant, previewContainerPat[0]);
      });

      room.on('trackAdded', function(track, participant) {
        //log(participant.identity + " added track: " + track.kind);
        var previewContainer = angular.element('#remote-media');
        attachTracks([track], previewContainer[0]);
      });
    });
  
    
  
  
    room.on('trackRemoved', function(track, participant) {
      //log(participant.identity + " removed track: " + track.kind);
      detachTracks([track]);
    });
  
  
    room.on('participantDisconnected', function(participant) {
      //log("Participant '" + participant.identity + "' left the room");
     // detachParticipantTracks(participant);
    });
  
    
    room.on('disconnected', function() {
     
      if (previewTracks) {
        previewTracks.forEach(function(track) {
          track.stop();
        });
        previewTracks = null;
      }
      
      room.participants.forEach(detachParticipantTracks);
      activeRoom = null;
    
    });
  }

function attachTracks(tracks, container) {
    tracks.forEach(function(track) {
      container.appendChild(track.attach());
    });
  }
  

  function attachParticipantTracks(participant, container, isLocal) {
    var tracks = getTracks(participant);
    attachTracks(tracks, container, isLocal);
 }
 
 function getTracks(participant) {
    return Array.from(participant.tracks.values()).filter(function (publication) {
       return publication.track;
    }).map(function (publication) {
       return publication.track;
    });
 }`

**Expected behavior:**

participant video should display

**Actual behavior:**

Cannot read property 'values' of undefined 
beacause participant.tracks is not present

**Software versions:**

 - [ ] Browser(s): chrome
 - [ ] Operating System: ubuntu
 - [ ] twilio-video.js: 2.0.0-beta6
 - [ ] Third-party libraries (e.g., Angular, React, etc.): Angular js
``

**Comment by manjeshbhargav** [comment]
Hi @12MIT0020 ,

Thanks for writing in with your issue. In this segment of your code:
```js
participant.tracks.forEach(function(participant) {
  var previewContainerPat = angular.element('#remote-media');
  console.log(previewContainerPat);
  attachParticipantTracks(participant, previewContainerPat[0]);
});
```
Since you are traversing the list of the Participant's tracks, `participant` is actually a Track and not a Participant. So, in order to play back media of existing Participants, rewrite the above code this way:
```js
room.on('participantConnected', function(participant) {
  var previewContainerPat = angular.element('#remote-media');
  console.log(previewContainerPat);
  attachParticipantTracks(participant, previewContainerPat[0]);
  ...
});
```

Hope this helps.

Thanks,

Manjesh Malavalli
JSDK Team

**Comment by 12MIT0020** [comment]
Hi @manjeshbhargav,thanks for commenting on issue. i have rewrite the code to:
`room.on('participantConnected', function(participant) {
      var previewContainerPat = angular.element('#remote-media');
      console.log(previewContainerPat);
      attachParticipantTracks(participant, previewContainerPat[0]);
    


      //When a Participant adds a Track, attach it to the DOM.
      room.on('trackAdded', function(track, participant) {
        //log(participant.identity + " added track: " + track.kind);
        var previewContainer = angular.element('#remote-media');
        attachTracks([track], previewContainer[0]);
      });
    });
  `
But i am getting issue in this piece of code : 
`  // Attach the Tracks to the DOM.
function attachTracks(tracks, container) {
    tracks.forEach(function(track) {
      container.appendChild(track.attach());
    });
  }
  
  // Attach the Participant's Tracks to the DOM.
  function attachParticipantTracks(participant, container, isLocal) {
    var tracks = getTracks(participant);
    attachTracks(tracks, container, isLocal);
 }
 
 function getTracks(participant) {
    return Array.from(participant.tracks.values()).filter(function (publication) {
       return publication.track;
    }).map(function (publication) {
       return publication.track;
    });
 }`

In getTracks function return publication.track which is null always.I checked in Participant.tracks.values it has RemotepublicationTrack object whose track property is null.so while control goes to attachTracks function it is attaching null to DOM.So no remote video  is not getting displayed .Please suggest.

**Comment by manjeshbhargav** [comment]
Hi @12MIT0020 ,

From your code snippet, it looks like you are using the 2.x beta version of the SDK. In this version, for any tracks you subscribe to after joining the Room, you have to listen to "trackSubscribed" and not "trackAdded":
```js
room.on('participantConnected', function(participant) {
  ...
  participant.on('trackSubscribed', function(track) {
    var previewContainer = angular.element('#remote-media');
    attachTracks([track], previewContainer[0]);
  });
  ...
});
```

Hope this helps.

Thanks,

Manjesh Malavalli
JSDK Team

**Comment by 12MIT0020** [comment]
Hi @manjeshbhargav  Thanks for support...it worked now..but still has a issue like whoever initiating the call that person can see and hear the video audio other side around is not happening(only getting preview of itshelf).Please suggest solution. Newer version of code is :
`function roomJoined(room) {
    console.log(room);
    window.room = activeRoom = room;
  
    var previewContainer = angular.element('#local-media');
    console.log(previewContainer);
   // if (!previewContainer.querySelector('video')) {
      attachParticipantTracks(room.localParticipant, previewContainer[0]);
   // }
  

    room.on('participantConnected', function(participant) {
      var previewContainerPat = angular.element('#remote-media');
      console.log(previewContainerPat);
      attachParticipantTracks(participant, previewContainerPat[0]);
    
      participant.on('trackSubscribed', function(track) {
        var previewContainer = angular.element('#remote-media');
        attachTracks([track], previewContainer[0]);
      });
    });
  
    
  
    room.on('trackRemoved', function(track, participant) {
      detachTracks([track]);
    });
  
    room.on('participantDisconnected', function(participant) {
      detachParticipantTracks(participant);
    });
 
    room.on('disconnected', function() {
      if (previewTracks) {
        previewTracks.forEach(function(track) {
          track.stop();
        });
        previewTracks = null;
      }
      room.participants.forEach(detachParticipantTracks);
      activeRoom = null;
    });
  }


function attachTracks(tracks, container) {
    tracks.forEach(function(track) {
      container.appendChild(track.attach());
    });
  }
  
  function attachParticipantTracks(participant, container, isLocal) {
    var tracks = getTracks(participant);
    attachTracks(tracks, container, isLocal);
 }
 
 function getTracks(participant) {
    return Array.from(participant.tracks.values()).filter(function (publication) {
       return publication.track;
    }).map(function (publication) {
       return publication.track;
    });
 }

function  detachParticipantTracks(participant) {
  var tracks = getTracks(participant);
  detachTracks(tracks);
}

function detachTracks(tracks) {
  for (var track of tracks) {
    const htmlElements = track.detach();
    for (var htmlElement of htmlElements) {
       htmlElement.remove();
     }
  }
}`

**Comment by Muzaffar-Munir** [comment]
Hi, getting the same issue in Vuejs in which the **initiator(first participant)** is getting a preview of all participants but the other participant is getting a preview of only participants entered later in a room. e.g "AA" is the **first Participant** and "BB" is the **second participant** and "CC" is **third Participant**. Now "AA" getting a preview of "BB" and "CC" but "BB" is only getting a preview of "CC" and "CC" is just getting his preview. I don't know where I mess up things
